### PR TITLE
Add support for server username/password auth

### PR DESCRIPTION
**Pull request by igfarm**
This change adds support for username/password authentication to the server via new server.set_user_manager(user_manager) function.  The user_manager function takes two arguments, namely a username and password, and returns a boolean indicating if access should be granted.

### CODE REVIEW

File: opcua/server/server.py [hunk 1]
```diff
@@ -641,3 +644,42 @@ def set_attribute_value(self, nodeid, datavalue, attr=ua.AttributeIds.Value):
         so it is a little faster
         """
         return self.iserver.set_attribute_value(nodeid, datavalue, attr)
+    
+    def set_user_manager(self, user_manager):
```
Comment (oroulet) [inline opcua/server/server.py]: I tend to think these methods should be part of InternalServer. They really implement new feature. The Server class is supposed to only be a high level helper class

File: opcua/server/server.py [hunk 2]
```diff
@@ -641,3 +644,42 @@ def set_attribute_value(self, nodeid, datavalue, attr=ua.AttributeIds.Value):
...
+        """
+        set up a function which that will check for authorize users. Input function takes username
+        and password as paramters and returns True of user is allowed access, False otherwise.
+        """
+        self.user_manager = user_manager
+        self.iserver.user_token_manager = self.user_token_manager
+
+    def user_token_manager(self, token):
```
Comment (oroulet) [inline opcua/server/server.py]: if this method does something (it seems it does ;-)) then I think its name should contain a verb explain what it does: check_user_token(token) for example. But feel free to propose something better

File: opcua/server/internal_server.py [hunk 3]
```diff
@@ -332,7 +333,10 @@ def activate_session(self, params):
         self.state = SessionState.Activated
         id_token = params.UserIdentityToken
         if isinstance(id_token, ua.UserNameIdentityToken):
-            if self.iserver.allow_remote_admin and id_token.UserName in ("admin", "Admin"):
+            if self.iserver.user_token_manager:
+                if self.iserver.user_token_manager(id_token) == False:
+                    raise utils.ServiceError(ua.StatusCodes.BadUserAccessDenied)
+            elif self.iserver.allow_remote_admin and id_token.UserName in ("admin", "Admin"):
```
Comment (oroulet) [inline opcua/server/internal_server.py]: should'nt we implement this default "admin" user using a default user_manager? instead of having a special case?

**Comment by oroulet** [comment]
thanks for the PR. This seems like a very simple solution the user/password issue, but please have a look at the comments.

File: opcua/server/internal_server.py (hunk 3 above)
Comment (igfarm) [reply opcua/server/internal_server.py]: That's a good idea.  I will implement that way.

File: opcua/server/server.py (hunk 1 above)
Comment (igfarm) [reply opcua/server/server.py]: The reason the user_token_manager function (or check_user_token) used in InternalServer is in Server is that it needs access to the Server.private_key to decode the token password. Any suggestion on how to gain access to this variable from InternalServer?

Comment (oroulet) [reply opcua/server/server.py]: It is probably wrong to store private_key in server. Can you move it to internal server?

**Comment by igfarm** [comment]
I moved the functions and variables  to the InternalServer as suggested. It looks cleaner now. The user_manager function is set up by calling InternalServer.set_user_manager().  The function itself requires an the extra isession parameter which is used to set up the  User.Admin type.  The default function is on InternalServer.default_user_manager()

File: opcua/server/internal_server.py [hunk 4]
```diff
@@ -269,7 +281,49 @@ def set_attribute_value(self, nodeid, datavalue, attr=ua.AttributeIds.Value):
...
+        """
+        self.user_manager = user_manager
 
+    def reset_user_manager(self):
+        """
+        reset user manager to default state 
+        """
+        self.user_manager = self.default_user_manager
+        
+    def default_user_manager(self, isession, username, password):
```
[the remark this answers is not in the document]
Comment (igfarm) [reply opcua/server/internal_server.py]: @oroulet, where would you like the function to live?

File: opcua/server/internal_server.py [hunk 5]
```diff
@@ -269,7 +281,46 @@ def set_attribute_value(self, nodeid, datavalue, attr=ua.AttributeIds.Value):
...
+            except Exception as exp:
+                self.logger.warning("Unable to decrypt password")
+                return False
+
+        # call user_manager if one is defined
+        if self.user_manager is not None:
+            return self.user_manager(self, isession, userName, passwd)
+
+        # If nothing function defined, do the default thing
+        if self.allow_remote_admin and userName in ("admin", "Admin"):
```
Comment (oroulet) [inline opcua/server/internal_server.py]: Why did you remove the default user manager function?I thought it was a good idea. It shows people how they can make they own

Comment (igfarm) [reply opcua/server/internal_server.py]: I see. I can add it back in.... where should this function live?

**Comment by igfarm** [comment]
I added the function back as an example. Thanks for your patience and guidance.

**Comment by igfarm** [comment]
Anything I can do to resolve the conflicts?

**Comment by oroulet** [comment]
@igfarm git pull --rebase origin master
you will get some conflicts to solve that should not be difficult
If you do have experience with conflict solving, remember to make a backup branch before so you do not lose any work

**Comment by oroulet** [comment]
something is wrong there is still a conflict so I just merged ans sqaushed everythig by hand

**Comment by oroulet** [comment]
thanks for the PR!
